src/tools/tools/process.fs

```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+            // Termniate all Wait thread. 
+            ThreadPoolWait.TryTerminateAll()
+    static member WaitForHandleWithCancellation (infoFunc: unit-> string) (handle:WaitHandle) (continuation:unit->unit) (unblockHandle:ManualResetEvent) (token:CancellationToken) =
+        token.Register( fun _ -> ThreadPoolWait.TryRemove handle |> ignore ) |> ignore 
+        ThreadPoolWait.WaitForHandle infoFunc handle continuation unblockHandle
+    static member WaitForHandle (infoFunc: unit-> string) (handle:WaitHandle) (continuation:unit->unit) (unblockHandle:ManualResetEvent) =
+        if handle.WaitOne(0) then 
+            continuation() 
+            if Utils.IsNotNull unblockHandle then 
+                // If there is an unblock handle, set it. 
+                unblockHandle.Set() |> ignore
+        else
+            /// Uniquely identify this async job and its resource removal. 
+            let jobObject = Object()
+            let rwh = ThreadPool.RegisterWaitForSingleObject( handle, new WaitOrTimerCallback(ThreadPoolWait.CallBack), (infoFunc,handle,continuation,unblockHandle,jobObject) , -1, true )
+            let handleCollection = ThreadPoolWait.WaitHandleCollection.GetOrAdd( handle, fun _ -> ConcurrentDictionary())
+            handleCollection.Item(jobObject) <- ( rwh, infoFunc, ref DateTime.MinValue.Ticks )
+    static member private CallBack( state: Object ) (timeout:bool) = 
+        if not timeout then
+            try
+                let infoFunc,handle,continuation,unblockHandle,jobObject = state :?> ((unit->string)*WaitHandle*(unit->unit)*ManualResetEvent*Object)
+                let bExistHandleCollection, handleCollection = ThreadPoolWait.WaitHandleCollection.TryGetValue( handle )
+                let bExist, tuple =  
+                    if bExistHandleCollection then 
+                        handleCollection.TryGetValue(jobObject)
+                    else
+                        false, Unchecked.defaultof<_>
+                let ticksRef = 
+                    if not bExist then 
+                        ref DateTime.MinValue.Ticks // Continue to execute 
+                    else
+                        let _, _, ticksRef = tuple
+                        ticksRef
+                let bCancelled = ((!ticksRef) = DateTime.MaxValue.Ticks)
+                if not bCancelled then 
+                    ticksRef := DateTime.UtcNow.Ticks
+                    continuation() 
+                    if Utils.IsNotNull unblockHandle  then 
+                        // If there is an unblock handle, set it. 
+                        unblockHandle.Set() |> ignore  
+                if bExist then 
+                    let rwh, _, _ = tuple 
+                    rwh.Unregister(null) |> ignore 
+                    handleCollection.TryRemove( jobObject ) |> ignore 
+                    if handleCollection.IsEmpty then
+                        ThreadPoolWait.WaitHandleCollection.TryRemove( handle ) |> ignore 
```

Comment:
This is not safe to do. It's possible this thread sees handleCollection.IsEmpty, then takes the branch, however, before it called "TryRemove", another thread can register a wait for the same handle.  This remove would cause the new handle to be removed, which should not happen.